File: solana-programs/scripts/airdrop/generate_merkle_tree.py
#!/usr/bin/env python3
"""
Merkle Tree Generation Script for KAMIYO Airdrop (Python Version)

This script generates a merkle tree from user allocations (points system data)
and creates proofs for each eligible wallet.

Usage:
    python generate_merkle_tree.py <input_csv> <output_json>

Input CSV format:
    wallet_address,points,allocation_kamiyo
    7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU,500,10000
    5Ghq4q5XLB4JUFy5YvGGaF5Vz8ZQjX8fJYy7TLZ8dZ8m,250,5000

Dependencies:
    pip install base58 pycryptodome
"""

import sys
import json
import csv
from typing import List, Dict, Tuple
from datetime import datetime
from Crypto.Hash import keccak
import base58


class MerkleTree:
    """Merkle tree generator for KAMIYO airdrop"""

    def __init__(self, allocations: List[Dict]):
        self.leaves = []
        self.tree = []

        # Create leaf nodes
        for alloc in allocations:
            amount_lamports = int(alloc['amount'] * 1_000_000_000)  # Convert to lamports
            leaf_hash = self._hash_leaf(alloc['address'], amount_lamports)

            self.leaves.append({
                'address': alloc['address'],
                'amount': amount_lamports,
                'amountKamiyo': alloc['amount'],
                'hash': leaf_hash
            })

        # Sort leaves for deterministic ordering
        self.leaves.sort(key=lambda x: x['hash'])

        # Build tree
        self._build_tree()

    def _hash_leaf(self, address: str, amount: int) -> str:
        """Hash a leaf node (wallet address + amount)"""
        try:
            # Decode base58 address to bytes (32 bytes)
            pubkey_bytes = base58.b58decode(address)
            if len(pubkey_bytes) != 32:
                raise ValueError(f"Invalid address length: {address}")

            # Convert amount to little-endian 8 bytes
            amount_bytes = amount.to_bytes(8, byteorder='little')

            # Concatenate and hash
            data = pubkey_bytes + amount_bytes
            k = keccak.new(digest_bits=256)
            k.update(data)
            return k.hexdigest()

        except Exception as e:
            raise ValueError(f"Invalid wallet address {address}: {e}")

    def _hash_pair(self, left: str, right: str) -> str:
        """Hash two nodes together"""
        # Sort for deterministic ordering
        first, second = (left, right) if left <= right else (right, left)

        # Concatenate hex strings and hash
        combined = bytes.fromhex(first + second)
        k = keccak.new(digest_bits=256)
        k.update(combined)
        return k.hexdigest()

    def _build_tree(self):
        """Build merkle tree from leaves"""
        current_level = [leaf['hash'] for leaf in self.leaves]
        self.tree.append(current_level[:])

        while len(current_level) > 1:
            next_level = []

            for i in range(0, len(current_level), 2):
                left = current_level[i]
                right = current_level[i + 1] if i + 1 < len(current_level) else left

                parent = self._hash_pair(left, right)
                next_level.append(parent)

            self.tree.append(next_level)
            current_level = next_level

    def get_root(self) -> str:
        """Get merkle root"""
        return self.tree[-1][0]

    def get_proof(self, address: str) -> List[str]:
        """Get proof for a specific wallet"""
        # Find leaf index
        leaf_index = None
        for i, leaf in enumerate(self.leaves):
            if leaf['address'] == address:
                leaf_index = i
                break

        if leaf_index is None:
            raise ValueError(f"Wallet {address} not found in tree")

        proof = []
        current_index = leaf_index

        # Traverse up the tree
        for level in range(len(self.tree) - 1):
            level_nodes = self.tree[level]
            is_right_node = current_index % 2 == 1
            sibling_index = current_index - 1 if is_right_node else current_index + 1

            if sibling_index < len(level_nodes):
                proof.append(level_nodes[sibling_index])

            current_index = current_index // 2

        return proof

    def get_all_proofs(self) -> List[Dict]:
        """Get proofs for all wallets"""
        proofs = []
        for leaf in self.leaves:
            proof = self.get_proof(leaf['address'])
            proofs.append({
                'address': leaf['address'],
                'amount': str(leaf['amount']),
                'amountKamiyo': leaf['amountKamiyo'],
                'proof': proof
            })
        return proofs

    def get_height(self) -> int:
        """Get tree height"""
        return len(self.tree)


def parse_allocations_csv(csv_path: str) -> List[Dict]:
    """Parse CSV file with allocations"""
    allocations = []

    with open(csv_path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            try:
                allocations.append({
                    'address': row['wallet_address'].strip(),
                    'points': int(row['points']),
                    'amount': float(row['allocation_kamiyo'])
                })
            except Exception as e:
                print(f"Warning: Skipping invalid row: {row} - {e}")

    return allocations


def main():
    if len(sys.argv) < 3:
        print('Usage: python generate_merkle_tree.py <input_csv> <output_json>')
        print('')
        print('Example:')
        print('  python generate_merkle_tree.py allocations.csv merkle-tree.json')
        sys.exit(1)

    input_path = sys.argv[1]
    output_path = sys.argv[2]

    print('🌳 KAMIYO Airdrop Merkle Tree Generator (Python)')
    print('')

    # Parse allocations
    print(f'📄 Reading allocations from: {input_path}')
    allocations = parse_allocations_csv(input_path)
    print(f'✅ Loaded {len(allocations)} allocations')

    # Validate allocations
    print('')
    print('🔍 Validating allocations...')
    total_tokens = sum(alloc['amount'] for alloc in allocations)
    print(f'   Total tokens: {total_tokens:,.0f} KAMIYO')

    max_allocation = max(alloc['amount'] for alloc in allocations)
    min_allocation = min(alloc['amount'] for alloc in allocations)
    print(f'   Range: {min_allocation:.0f} - {max_allocation:.0f} KAMIYO')

    # Check caps
    exceeds_max = [a for a in allocations if a['amount'] > 10_000]
    if exceeds_max:
        print(f'⚠️  Warning: {len(exceeds_max)} allocations exceed 10,000 KAMIYO cap')

    below_min = [a for a in allocations if a['amount'] < 100]
    if below_min:
        print(f'⚠️  Warning: {len(below_min)} allocations below 100 KAMIYO minimum')

    # Generate merkle tree
    print('')
    print('🌲 Generating merkle tree...')
    tree = MerkleTree(allocations)
    root = tree.get_root()
    print(f'✅ Merkle root: {root}')
    print(f'   Tree height: {tree.get_height()} levels')

    # Generate proofs
    print('')
    print('🔑 Generating proofs for all wallets...')
    proofs = tree.get_all_proofs()
    print(f'✅ Generated {len(proofs)} proofs')

    # Create output
    output = {
        'merkleRoot': root,
        'totalAllocations': len(allocations),
        'totalTokens': str(int(total_tokens * 1_000_000_000)),
        'totalTokensKamiyo': total_tokens,
        'treeHeight': tree.get_height(),
        'leaves': tree.leaves,
        'proofs': proofs,
        'generatedAt': datetime.utcnow().isoformat() + 'Z'
    }

    # Write output
    print('')
    print(f'💾 Writing output to: {output_path}')
    with open(output_path, 'w') as f:
        json.dump(output, f, indent=2)
    print('✅ Done!')

    # Summary
    print('')
    print('📊 Summary:')
    print(f'   Merkle Root: {root}')
    print(f'   Total Allocations: {len(allocations)}')
    print(f'   Total Tokens: {total_tokens:,.0f} KAMIYO')
    print(f'   Tree Height: {tree.get_height()} levels')
    print('')
    print('🚀 Next steps:')
    print('   1. Review the generated merkle tree JSON')
    print('   2. Initialize airdrop program with merkle root')
    print('   3. Fund vault with tokens')
    print('   4. Users can claim using their proof from the JSON file')


if __name__ == '__main__':
    main()

File: solana-programs/scripts/airdrop/generate-merkle-tree.ts
/**
 * Merkle Tree Generation Script for KAMIYO Airdrop
 *
 * This script generates a merkle tree from user allocations (points system data)
 * and creates proofs for each eligible wallet.
 *
 * Usage:
 *   ts-node generate-merkle-tree.ts <input_csv> <output_json>
 *
 * Input CSV format:
 *   wallet_address,points,allocation_kamiyo
 *   7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU,500,10000
 *   5Ghq4q5XLB4JUFy5YvGGaF5Vz8ZQjX8fJYy7TLZ8dZ8m,250,5000
 */

import { PublicKey } from '@solana/web3.js';
import { keccak_256 } from 'js-sha3';
import * as fs from 'fs';
import * as path from 'path';

interface Allocation {
    address: string;
    points: number;
    amount: number; // In KAMIYO tokens (not lamports)
}

interface LeafData {
    address: string;
    amount: number; // In lamports
    amountKamiyo: number; // In KAMIYO for display
    hash: string;
}

interface ProofData {
    address: string;
    amount: string; // String to preserve precision
    amountKamiyo: number;
    proof: string[];
}

interface MerkleTreeOutput {
    merkleRoot: string;
    totalAllocations: number;
    totalTokens: string; // In lamports
    totalTokensKamiyo: number;
    treeHeight: number;
    leaves: LeafData[];
    proofs: ProofData[];
    generatedAt: string;
}

class MerkleTree {
    private leaves: LeafData[] = [];
    private tree: string[][] = [];

    constructor(allocations: Allocation[]) {
        // Create leaf nodes
        this.leaves = allocations.map(alloc => {
            const amountLamports = BigInt(alloc.amount) * BigInt(1_000_000_000); // Convert to lamports
            const leaf = this.hashLeaf(alloc.address, amountLamports);

            return {
                address: alloc.address,
                amount: Number(amountLamports),
                amountKamiyo: alloc.amount,
                hash: leaf,
            };
        });

        // Sort leaves for deterministic ordering
        this.leaves.sort((a, b) => a.hash.localeCompare(b.hash));

        // Build tree bottom-up
        this.buildTree();
    }

    /**
     * Hash a leaf node (wallet address + amount)
     */
    private hashLeaf(address: string, amount: bigint): string {
        try {
            // Validate address
            const pubkey = new PublicKey(address);

            // Create buffer: [pubkey_bytes (32) || amount_le_bytes (8)]
            const buffer = Buffer.alloc(40);
            pubkey.toBuffer().copy(buffer, 0);

            // Write amount as little-endian 64-bit
            const amountBuffer = Buffer.alloc(8);
            amountBuffer.writeBigUInt64LE(amount);
            amountBuffer.copy(buffer, 32);

            // Hash with keccak256
            const hash = keccak_256(buffer);
            return hash;
        } catch (err) {
            throw new Error(`Invalid wallet address: ${address}`);
        }
    }

    /**
     * Hash two nodes together
     */
    private hashPair(left: string, right: string): string {
        // Sort for deterministic ordering
        const [first, second] = left <= right ? [left, right] : [right, left];

        const combined = Buffer.from(first + second, 'hex');
        return keccak_256(combined);
    }

    /**
     * Build merkle tree from leaves
     */
    private buildTree(): void {
        let currentLevel = this.leaves.map(leaf => leaf.hash);
        this.tree.push([...currentLevel]);

        while (currentLevel.length > 1) {
            const nextLevel: string[] = [];

            for (let i = 0; i < currentLevel.length; i += 2) {
                const left = currentLevel[i];
                const right = i + 1 < currentLevel.length ? currentLevel[i + 1] : left;

                const parent = this.hashPair(left, right);
                nextLevel.push(parent);
            }

            this.tree.push(nextLevel);
            currentLevel = nextLevel;
        }
    }

    /**
     * Get merkle root
     */
    public getRoot(): string {
        return this.tree[this.tree.length - 1][0];
    }

    /**
     * Get proof for a specific wallet
     */
    public getProof(address: string): string[] {
        // Find leaf index
        const leafIndex = this.leaves.findIndex(leaf => leaf.address === address);
        if (leafIndex === -1) {
            throw new Error(`Wallet ${address} not found in tree`);
        }

        const proof: string[] = [];
        let currentIndex = leafIndex;

        // Traverse up the tree
        for (let level = 0; level < this.tree.length - 1; level++) {
            const levelNodes = this.tree[level];
            const isRightNode = currentIndex % 2 === 1;
            const siblingIndex = isRightNode ? currentIndex - 1 : currentIndex + 1;

            if (siblingIndex < levelNodes.length) {
                proof.push(levelNodes[siblingIndex]);
            }

            currentIndex = Math.floor(currentIndex / 2);
        }

        return proof;
    }

    /**
     * Get all proofs
     */
    public getAllProofs(): ProofData[] {
        return this.leaves.map(leaf => ({
            address: leaf.address,
            amount: leaf.amount.toString(),
            amountKamiyo: leaf.amountKamiyo,
            proof: this.getProof(leaf.address),
        }));
    }

    /**
     * Get tree height
     */
    public getHeight(): number {
        return this.tree.length;
    }

    /**
     * Get leaves
     */
    public getLeaves(): LeafData[] {
        return this.leaves;
    }
}

/**
 * Parse CSV file with allocations
 */
function parseAllocationsCSV(csvPath: string): Allocation[] {
    const content = fs.readFileSync(csvPath, 'utf-8');
    const lines = content.trim().split('\n');

    // Skip header
    const dataLines = lines.slice(1);

    const allocations: Allocation[] = [];

    for (const line of dataLines) {
        const [address, points, amount] = line.split(',');

        if (!address || !points || !amount) {
            console.warn(`Skipping invalid line: ${line}`);
            continue;
        }

        allocations.push({
            address: address.trim(),
            points: parseInt(points.trim()),
            amount: parseFloat(amount.trim()),
        });
    }

    return allocations;
}

/**
 * Main execution
 */
function main() {
    const args = process.argv.slice(2);

    if (args.length < 2) {
        console.error('Usage: ts-node generate-merkle-tree.ts <input_csv> <output_json>');
        console.error('');
        console.error('Example:');
        console.error('  ts-node generate-merkle-tree.ts allocations.csv merkle-tree.json');
        process.exit(1);
    }

    const [inputPath, outputPath] = args;

    console.log('🌳 KAMIYO Airdrop Merkle Tree Generator');
    console.log('');

    // Parse allocations
    console.log(`📄 Reading allocations from: ${inputPath}`);
    const allocations = parseAllocationsCSV(inputPath);
    console.log(`✅ Loaded ${allocations.length} allocations`);

    // Validate allocations
    console.log('');
    console.log('🔍 Validating allocations...');
    const totalTokens = allocations.reduce((sum, alloc) => sum + alloc.amount, 0);
    console.log(`   Total tokens: ${totalTokens.toLocaleString()} KAMIYO`);

    const maxAllocation = Math.max(...allocations.map(a => a.amount));
    const minAllocation = Math.min(...allocations.map(a => a.amount));
    console.log(`   Range: ${minAllocation} - ${maxAllocation} KAMIYO`);

    // Check caps
    const exceedsMax = allocations.filter(a => a.amount > 10_000);
    if (exceedsMax.length > 0) {
        console.warn(`⚠️  Warning: ${exceedsMax.length} allocations exceed 10,000 KAMIYO cap`);
    }

    const belowMin = allocations.filter(a => a.amount < 100);
    if (belowMin.length > 0) {
        console.warn(`⚠️  Warning: ${belowMin.length} allocations below 100 KAMIYO minimum`);
    }

    // Generate merkle tree
    console.log('');
    console.log('🌲 Generating merkle tree...');
    const tree = new MerkleTree(allocations);
    const root = tree.getRoot();
    console.log(`✅ Merkle root: ${root}`);
    console.log(`   Tree height: ${tree.getHeight()} levels`);

    // Generate proofs
    console.log('');
    console.log('🔑 Generating proofs for all wallets...');
    const proofs = tree.getAllProofs();
    console.log(`✅ Generated ${proofs.length} proofs`);

    // Create output
    const output: MerkleTreeOutput = {
        merkleRoot: root,
        totalAllocations: allocations.length,
        totalTokens: (BigInt(totalTokens) * BigInt(1_000_000_000)).toString(),
        totalTokensKamiyo: totalTokens,
        treeHeight: tree.getHeight(),
        leaves: tree.getLeaves(),
        proofs: proofs,
        generatedAt: new Date().toISOString(),
    };

    // Write output
    console.log('');
    console.log(`💾 Writing output to: ${outputPath}`);
    fs.writeFileSync(outputPath, JSON.stringify(output, null, 2));
    console.log('✅ Done!');

    // Summary
    console.log('');
    console.log('📊 Summary:');
    console.log(`   Merkle Root: ${root}`);
    console.log(`   Total Allocations: ${allocations.length}`);
    console.log(`   Total Tokens: ${totalTokens.toLocaleString()} KAMIYO`);
    console.log(`   Tree Height: ${tree.getHeight()} levels`);
    console.log('');
    console.log('🚀 Next steps:');
    console.log('   1. Review the generated merkle tree JSON');
    console.log('   2. Initialize airdrop program with merkle root');
    console.log('   3. Fund vault with tokens');
    console.log('   4. Users can claim using their proof from the JSON file');
}

// Run if called directly
if (require.main === module) {
    main();
}

export { MerkleTree, Allocation, ProofData, MerkleTreeOutput };
